### PR TITLE
msgpack-cxx 4.0.1 (new formula)

### PR DESCRIPTION
**Pull request by chenrui333**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [x] Have you ensured that your commits follow the [commit style guide](https://docs.brew.sh/Formula-Cookbook#commit)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [ ] Does your build pass `brew audit --strict <formula>` (after doing `brew install --build-from-source <formula>`)? If this is a new formula, does it pass `brew audit --new <formula>`?

---

it failed to pass the audit check

```
msgpack-cpp:
  * HEAD: Use `branch: "master"` to specify the default branch
```

We might want to have a allowlist for branch audit check.

### CODE REVIEW

File: Formula/msgpack-cpp.rb [hunk 1]
```diff
@@ -0,0 +1,54 @@
+class MsgpackCpp < Formula
+  desc "MessagePack implementation for C++ / msgpack.org[C++]"
+  homepage "https://msgpack.org/"
+  url "https://github.com/msgpack/msgpack-c/releases/download/cpp-4.0.1/msgpack-cxx-4.0.1.tar.gz"
+  sha256 "acba76ebc3ff9af2fd2649f319b25cd2387c66f75ae74a7a652fe3386e45d498"
+  license "BSL-1.0"
+  head "https://github.com/msgpack/msgpack-c.git", branch: "cpp_master"
+
+  depends_on "boost" => :build
```
Comment (carlocab) [inline Formula/msgpack-cpp.rb:9]: I think this is needed at runtime.

File: Formula/msgpack.rb [hunk 2]
```diff
@@ -17,13 +17,15 @@ class Msgpack < Formula
 
   depends_on "cmake" => :build
 
+  conflicts_with "msgpack-cpp", because: "both install `msgpack` binaries"
```
Comment (carlocab) [inline Formula/msgpack.rb:20]:
```suggestion
```
This is not true. `msgpack` does not ship any binaries.
```
❯ brew ls -l msgpack
/usr/local/Cellar/msgpack/3.3.0/include/msgpack/ (737 files)
/usr/local/Cellar/msgpack/3.3.0/include/ (2 files)
/usr/local/Cellar/msgpack/3.3.0/lib/libmsgpackc.2.0.0.dylib
/usr/local/Cellar/msgpack/3.3.0/lib/cmake/ (4 files)
/usr/local/Cellar/msgpack/3.3.0/lib/pkgconfig/msgpack.pc
/usr/local/Cellar/msgpack/3.3.0/lib/ (3 other files)
```

File: Formula/msgpack-cpp.rb [hunk 3]
```diff
@@ -0,0 +1,54 @@
+class MsgpackCpp < Formula
```
Comment (carlocab) [inline Formula/msgpack-cpp.rb:1]: Should we call this `msgpack-cxx`?

File: Formula/msgpack-cpp.rb [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+  homepage "https://msgpack.org/"
+  url "https://github.com/msgpack/msgpack-c/releases/download/cpp-4.0.1/msgpack-cxx-4.0.1.tar.gz"
+  sha256 "acba76ebc3ff9af2fd2649f319b25cd2387c66f75ae74a7a652fe3386e45d498"
+  license "BSL-1.0"
+  head "https://github.com/msgpack/msgpack-c.git", branch: "cpp_master"
+
+  depends_on "boost" => :build
+  depends_on "cmake" => :build
+
+  conflicts_with "msgpack", because: "both install `msgpack` binaries"
```
Comment (carlocab) [inline Formula/msgpack-cpp.rb:12]: I think the conflict comes from both installing `include/msgpack`. Perhaps we can rename the one here to `include/msgpack-cxx`?
```ruby
include.install include/"msgpack" => "msgpack-cxx"
```
That we we won't need the conflict.

File: Formula/msgpack-cpp.rb (hunk 3 above)
Comment (chenrui333) [reply Formula/msgpack-cpp.rb:1]: I saw most of the formulae used `cpp` part of the names, also the head branch is also with `cpp_master`

File: Formula/msgpack-cpp.rb (hunk 4 above)
Comment (chenrui333) [reply Formula/msgpack-cpp.rb:12]: ok, let's do that.

**Comment by chenrui333** [comment]
Considering most of the formulae are with name of `cpp` as well as the head branch is with `cpp_master`, we should use `msgpack-cpp`, but the artifact is with `msgpack-cxx` though.

File: Formula/msgpack-cpp.rb (hunk 1 above)
Comment (chenrui333) [reply Formula/msgpack-cpp.rb:9]: that sounds about right!

File: Formula/msgpack-cpp.rb (hunk 4 above)
Comment (chenrui333) [reply Formula/msgpack-cpp.rb]: The out-of-the-box example does not work with this change.

File: Formula/msgpack.rb [hunk 5]
```diff
@@ -4,7 +4,7 @@ class Msgpack < Formula
   url "https://github.com/msgpack/msgpack-c/releases/download/cpp-3.3.0/msgpack-3.3.0.tar.gz"
   sha256 "6e114d12a5ddb8cb11f669f83f32246e484a8addd0ce93f274996f1941c1f07b"
   license "BSL-1.0"
-  head "https://github.com/msgpack/msgpack-c.git"
+  head "https://github.com/msgpack/msgpack-c.git", branch: "c_master"
```
Comment (carlocab) [inline Formula/msgpack.rb:7]: Let's do this in a separate PR so we don't have to rebuild bottles for this and throw away the `high_sierra` bottle.

File: Formula/msgpack-cxx.rb [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+    # fix include conflicts between msgpack and msgpack-cxx
+    include.install include/"msgpack" => "msgpack-cxx"
+    include.install include/"msgpack.hpp" => "msgpack-cxx.hpp"
+    inreplace include/"msgpack-cxx.hpp", "\"msgpack/", "\"msgpack-cxx/"
+  end
+
+  test do
+    # Reference: https://github.com/msgpack/msgpack-c/blob/cpp_master/QUICKSTART-CPP.md
+    (testpath/"test.cpp").write <<~EOS
+      #include <msgpack.hpp>
```
Comment (carlocab) [inline Formula/msgpack-cxx.rb:25]:
```suggestion
      #include <msgpack-cxx.hpp>
```

Comment (chenrui333) [reply Formula/msgpack-cxx.rb:25]: With that change, this is the error in my local
```
In file included from test.cpp:1:
In file included from /usr/local/Cellar/msgpack-cxx/4.0.1/include/msgpack-cxx.hpp:10:
/usr/local/Cellar/msgpack-cxx/4.0.1/include/msgpack-cxx/object.hpp:13:10: fatal error: 'msgpack/object_decl.hpp' file not found
#include "msgpack/object_decl.hpp"
         ^~~~~~~~~~~~~~~~~~~~~~~~~
1 error generated.
Error: msgpack-cxx: failed
```

File: Formula/msgpack.rb (hunk 5 above)
Comment (chenrui333) [reply Formula/msgpack.rb:7]: Sounds good.

Comment (chenrui333) [reply Formula/msgpack.rb]: Moved into #84264

File: Formula/msgpack-cxx.rb [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+  depends_on "boost"
+
+  def install
+    system "cmake", ".", *std_cmake_args
+    system "make", "install"
+
+    # fix include conflicts between msgpack and msgpack-cxx
+    include.install include/"msgpack" => "msgpack-cxx"
+    include.install include/"msgpack.hpp" => "msgpack-cxx.hpp"
+    inreplace include/"msgpack-cxx.hpp", "\"msgpack/", "\"msgpack-cxx/"
```
Comment (carlocab) [inline Formula/msgpack-cxx.rb:19]: Oops. Looks like renaming didn't work. Let's drop the renaming, and add the conflict in a follow-up PR so we can keep the `msgpack` bottle for High Sierra.

Comment (carlocab) [reply Formula/msgpack-cxx.rb:19]: Yup, don't add `conflicts_with` for now. Let's add it right after this is merged.

Comment (chenrui333) [reply Formula/msgpack-cxx.rb:19]: I see your poiint.

**Review by carlocab — APPROVED** [review]
Failures on ARM and Linux are audit failures for the `head` branch, but those are wrong. Assuming no build failures on Intel macOS, this should be good to merge.

**Pull request closed without merging**